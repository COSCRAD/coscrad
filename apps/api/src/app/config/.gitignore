.env
production.env
test.env
development.env

