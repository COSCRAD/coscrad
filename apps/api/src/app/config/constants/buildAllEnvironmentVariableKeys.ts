/**
 * We build this newly each time to avoid shared references \ side-effects.
 */
export default () => [
  'NODE_PORT',
  'ARANGO_DB_HOST_SCHEME',
  'ARANGO_DB_HOST_DOMAIN',
  'ARANGO_DB_HOST_PORT',
  'ARANGO_DB_ROOT_PASSWORD',
  'ARANGO_DB_USER',
  'ARANGO_DB_USER_PASSWORD',
  'ARANGO_DB_NAME',
  'AUTH0_ISSUER_URL',
  'AUTH0_AUDIENCE',
];
