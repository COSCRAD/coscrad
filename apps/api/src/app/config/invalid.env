# This is for tests only! (TODO generate and remove during test)
NODE_PORT='ABCDEFU' # invalid
ARANGO_DB_HOST_SCHEME=https
ARANGO_DB_HOST_DOMAIN=arango.mydomain.ca
ARANGO_DB_HOST_PORT=443
ARANGO_DB_ROOT_PASSWORD=super-confidential
ARANGO_DB_USER=5 # invalid
# ARANGO_DB_USER_PASSWORD=youllNEVERguess1!
ARANGO_DB_NAME=bogusdbname
ARANGO_QUERY_DB_NAME=sample_query_db
AUTH0_ISSUER_URL=https://foo.auth0.com/
AUTH0_AUDIENCE=https://foo.mydomain.ca
GLOBAL_PREFIX=api
