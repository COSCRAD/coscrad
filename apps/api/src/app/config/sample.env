NODE_PORT=1234
ARANGO_DB_HOST_SCHEME=https
ARANGO_DB_HOST_DOMAIN=arango.mydomain.ca
ARANGO_DB_HOST_PORT=443
ARANGO_DB_ROOT_PASSWORD=super-confidential
ARANGO_DB_USER=username
ARANGO_DB_USER_PASSWORD=youllNEVERguess1!
ARANGO_DB_NAME=bogusdbname
AUTH0_ISSUER_URL=https://foo.auth0.com/
AUTH0_AUDIENCE=https://bar.mydomain.ca