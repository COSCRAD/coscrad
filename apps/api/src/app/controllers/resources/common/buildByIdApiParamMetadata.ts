export default () => ({
    name: 'id',
    required: true,
    example: '2',
});
