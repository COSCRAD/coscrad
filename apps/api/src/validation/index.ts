export * from './dynamic-data-type-finder.service';
export * from './dynamic-data-type.module';
