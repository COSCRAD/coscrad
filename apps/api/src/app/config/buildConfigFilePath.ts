import { existsSync } from 'fs';
import { InternalError } from '../../lib/errors/InternalError';
import { Environment } from './constants/Environment';

const getTargetDirectoryForEnvironment = (environment: Environment): string => {
    if ([Environment.production, Environment.staging, Environment.e2e].includes(environment))
        return '/';

    return `/apps/api/src/app/config/`;
};

/**
 * It is the responsibility of the client to validate the environment prefix.
 */
export default (envFilePrefix: string): string => {
    // TODO We may want to throw here to fail faster when an invalid env is used

    /**
     * `envFilePrefix` is usually linked to node_env and of type `Environment`,
     * but we override this in some tests to use dummy filenames. For that reason,
     * we assume only that this is a string.
     */
    const baseDir = getTargetDirectoryForEnvironment(envFilePrefix as Environment);

    const path = `${process.cwd()}${baseDir}${envFilePrefix}.env`;

    if (!existsSync(path)) {
        throw new InternalError(`Invalid .env file path: ${path}`);
    }
    return path;
};
