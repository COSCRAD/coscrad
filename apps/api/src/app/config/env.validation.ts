import {
    ExternalEnum,
    getCoscradDataSchema,
    NonEmptyString,
    PositiveInteger,
    String,
    URL,
    validateCoscradModelInstance,
} from '@coscrad/data-types';
import { plainToInstance } from 'class-transformer';
import { Environment } from './constants/environment';
import { Scheme } from './constants/Scheme';

export class EnvironmentVariables {
    @ExternalEnum(
        {
            enumLabel: 'Node Environment',
            enumName: `Environment`,
            labelsAndValues: Object.values(Environment).map((envName) => ({
                label: envName,
                value: envName,
            })),
        },
        {
            label: 'Node Environment',
            description: 'is the server running in production, staging, testing?',
        }
    )
    NODE_ENV: Environment;

    @PositiveInteger({
        label: 'Node Port',
        description: 'the port on which Node (Express) will listen for requests',
    })
    NODE_PORT: number;

    // TODO[https://www.pivotaltracker.com/story/show/181530113] tighten up these rules with custom validators
    @String({
        label: 'Arango DB host scheme',
        description: 'http or https?',
    })
    ARANGO_DB_HOST_SCHEME: Scheme;

    @String({
        label: 'Arango DB host domain',
        description: "the domain part of Arango's rest endpoint",
    })
    ARANGO_DB_HOST_DOMAIN: string;

    @PositiveInteger({
        label: 'Arango Port',
        description: "the port Arango's rest server is listening on",
    })
    ARANGO_DB_HOST_PORT: number;

    @NonEmptyString({
        label: 'Arango DB root password',
        description: 'the password for the root user of your Arango instance',
    })
    ARANGO_DB_ROOT_PASSWORD: string;

    @NonEmptyString({
        label: 'Arango DB username',
        description: 'the username for the non-root system user of your Arango instance',
    })
    ARANGO_DB_USER: string;

    @NonEmptyString({
        label: 'Arango DB user password',
        description: 'the password for the non-root system user of your Arango instance',
    })
    ARANGO_DB_USER_PASSWORD: string;

    @NonEmptyString({
        label: 'Arango DB name',
        description: 'the name of the database your application will connect to',
    })
    ARANGO_DB_NAME: string;

    @URL({
        label: 'Auth0 Issuer URL',
        description: 'the issuer URL from Auth0',
    })
    AUTH0_ISSUER_URL: string;

    @URL({
        label: 'Auth0 Audience',
        description: 'the audience of the Auth0 App',
    })
    AUTH0_AUDIENCE: string;

    @URL({
        label: 'base API url',
        description:
            'the protocol and full base url (TLD and SLD) and port number, e.g., "https://www.myorg.com:8345"',
    })
    BASE_URL: string;

    @NonEmptyString({
        label: 'Global API Prefix',
        description: 'the prefix that will come before all endpoints in your API',
    })
    GLOBAL_PREFIX: string;

    // TODO Add Boolean Decorator
    @NonEmptyString({
        label: 'Games Endpoint Feature Flag',
        description: 'whether or not to enable the legacy games data endpoint',
    })
    SHOULD_ENABLE_LEGACY_GAMES_ENDPOINT: string;

    // TODO Can we add a parsing layer before validation? That way, this will be `string[]` by the time it's available in the `ConfigService`
    @NonEmptyString({
        label: 'allowed client origins',
        description:
            'Comma separated list. If non-empty, this list will be use to configure CORS for provided client URLs',
        isOptional: true,
    })
    ALLOWED_CLIENT_ORIGINS: string;
}

// Do we have a test of this?
export const validate = (config: Record<string, unknown>): EnvironmentVariables => {
    const configToValidate = plainToInstance(EnvironmentVariables, config, {
        enableImplicitConversion: true,
    });

    const errors = validateCoscradModelInstance(
        getCoscradDataSchema(EnvironmentVariables),
        configToValidate,
        {
            forbidUnknownValues: false,
        }
    );

    if (errors.length > 0) {
        throw new Error(errors.toString());
    }

    return configToValidate;
};
