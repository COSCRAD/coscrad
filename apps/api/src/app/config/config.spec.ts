import { INestApplication } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { Test } from '@nestjs/testing';
import buildConfigFilePath from '../config/buildConfigFilePath';
import { validate } from '../config/env.validation';
import buildAllEnvironmentVariableKeys from './constants/buildAllEnvironmentVariableKeys';
import removeAllCustomEntironmentVariables from './__tests__/utilities/removeAllCustomEnvironmentVariables';

/**
 * The purpose of the base '/' endpoint is to give us a sanity check that the
 * server is up and running properly. The purpose of this test is to give some
 * guidance in how to setup more complex e2e tests of our api.
 */

const createAppModule = (configFileName: string) =>
  Test.createTestingModule({
    imports: [
      ConfigModule.forRoot({
        isGlobal: true,
        /**
         * We intentionally read the sample .env here to avoid snapshotting
         * actual sensitive data. The only constraint is that the values
         * in `sample.env` need to have the corredt types as specified in
         * `env.validation.ts`.
         */
        envFilePath: buildConfigFilePath(configFileName),
        cache: false,
        validate,
      }),
    ],
  }).compile();

describe('ConfigService', () => {
  let app: INestApplication;

  beforeEach(async () => {
    /**
     * HACK We have experienced annoying side-effect issues with the way the
     * built-in ConfigService reads our `.env` files. We have a separate
     * `${environment}.env` file for each unique environment. However,
     * `process.env` apparently takes priority over these locally defined files.
     * Forcing the local values to take precedence in the test environment has
     * been an exercise in frustration.
     *
     * https://github.com/nestjs/config/issues/168
     *
     * https://github.com/nestjs/config/issues/168
     */
    removeAllCustomEntironmentVariables();
  });

  const expectedDataInResponse = {
    message: 'Welcome to the COSCRAD API!',
  };

  describe('the environment', () => {
    it('should be test', () => {
      expect(process.env.NODE_ENV).toBe('test');
    });
  });

  describe('when all environment variables are valid', () => {
    beforeAll(async () => {
      const moduleRef = await createAppModule('sample');

      app = moduleRef.createNestApplication();
      await app.init();
    });
    it('should set the variables appropriately', () => {
      const configService = app.get(ConfigService);

      const config = buildAllEnvironmentVariableKeys().reduce(
        (accumulated, key) => ({
          ...accumulated,
          [key]: configService.get(key),
        }),
        {}
      );

      expect(config).toMatchSnapshot();
    });
  });

  describe('when the .env contains invalid variable declarations', () => {
    const attemptToCreateAppModule = () => createAppModule('invalid');

    // We might want to check that the promise rejects here instead
    it('should throw', () => {
      expect(attemptToCreateAppModule).toThrow();

      let configErrorMessage;

      try {
        attemptToCreateAppModule();
      } catch (error) {
        configErrorMessage = error;
      }

      expect(configErrorMessage).toMatchSnapshot();
    });
  });

  afterEach(() => {
    removeAllCustomEntironmentVariables();
  });

  afterAll(async () => {
    await app.close();
  });
});
