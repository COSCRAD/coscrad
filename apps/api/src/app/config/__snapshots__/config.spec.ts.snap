// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigService when NODE_ENV is not inherited from the process should throw 1`] = `[Error: Error: Property NODE_ENV has failed the validation constraint: is enum Environment]`;

exports[`ConfigService when all environment variables are valid the config service when reading environment variables should obtain the correct values 1`] = `
{
  "ARANGO_DB_HOST_DOMAIN": "localhost",
  "ARANGO_DB_HOST_PORT": 9595,
  "ARANGO_DB_HOST_SCHEME": "http",
  "ARANGO_DB_NAME": "sampledb",
  "ARANGO_DB_ROOT_PASSWORD": "super-confidential",
  "ARANGO_DB_USER": "username",
  "ARANGO_DB_USER_PASSWORD": "youllNEVERguess1!",
  "AUTH0_AUDIENCE": "https://bar.mydomain.ca",
  "AUTH0_ISSUER_URL": "https://foo.auth0.com/",
  "NODE_PORT": 1234,
}
`;

exports[`ConfigService when the .env contains invalid variable declarations should throw 1`] = `[Error: Error: NODE_PORT: NaN has failed validation constraint: integer,Error: NODE_PORT: NaN has failed validation constraint: positive number,Error: ARANGO_DB_USER_PASSWORD: undefined has failed validation constraint: non-empty string,Error: ARANGO_DB_USER_PASSWORD: undefined has failed validation constraint: defined value]`;
